ambry-router/src/main/java/com.github.ambry.router/DefaultGCMCryptoService.java

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.config.CryptoServiceConfig;
+import com.github.ambry.messageformat.MessageFormatErrorCodes;
+import com.github.ambry.messageformat.MessageFormatException;
+import com.github.ambry.utils.ByteBufferInputStream;
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.security.Security;
+import javax.crypto.Cipher;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.util.encoders.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Default {@link CryptoService} which is capable of encrypting or decrypting bytes based on the given key.
+ * This implementation uses GCM for encryption and decryption
+ */
+class DefaultGCMCryptoService implements CryptoService<SecretKeySpec> {
+
+  private static final int Version_Field_Size_In_Bytes = Short.BYTES;
+  private static final short KeyRecord_Version_V1 = 1;
+  public static final short IVRecord_Version_V1 = 1;
+  private static final String GCM_CRYPTO_INSTANCE = "AES/GCM/NoPadding";
+  private static final SecureRandom random = new SecureRandom();
+  private final int ivValSize;
+  private final int keySize;
+  private final CryptoServiceConfig config;
+
+  private static final Logger logger = LoggerFactory.getLogger(DefaultGCMCryptoService.class);
+
+  DefaultGCMCryptoService(CryptoServiceConfig cryptoServiceConfig) {
+    config = cryptoServiceConfig;
+    ivValSize = cryptoServiceConfig.cryptoServiceIvSizeInBytes;
+    keySize = cryptoServiceConfig.cryptoServiceKeySizeInChars;
+    Security.addProvider(new BouncyCastleProvider());
+    if (!config.cryptoServiceEncryptionDecryptionMode.equals("GCM")) {
+      throw new IllegalArgumentException(
+          "Unrecognized Encryption Decryption Mode " + config.cryptoServiceEncryptionDecryptionMode);
+    }
+  }
+
+  @Override
+  public ByteBuffer encrypt(ByteBuffer toEncrypt, SecretKeySpec key) throws GeneralSecurityException {
+    try {
+      Cipher encrypter = Cipher.getInstance(GCM_CRYPTO_INSTANCE, "BC");
```

Comment:
Is getting a cipher instance an expensive operation? You might be able to use a threadlocal variable to cache them, but we probably shouldn't worry about doing this now.

Reply:
But we need to initialize the Cipher on every encrypt call and hence couldn't help much.